Make input contiguous for DTensor reduce scatter to fix the incorrect numerical values

yoyoyocmu opened the pull request:
Summary:
This change is to make the input tensor contiguous for DTensor reduce scatter in the case no padding is needed.

There's no exception thrown during training, but we ran into numerical value correctness issue without the change.

Test Plan:
**CI**
CI test

**WHEN model test**:
- Verified loss for each iteration within the expected range.
- Verified NE on-par with this change with 4B training data.

Differential Revision: D52170822




cc @mrshenli @pritamdamania87 @zhaojuanmao @satgera @rohan-varma @gqchen @aazzolini @osalpekar @jiayisuse @H-Huang @kwen2501 @awgu @penguinwu @fegin @XilunWu @wanchaol @fduwjj @wz337 @tianyu-l @wconstab @yf225

wanchaol commented on torch/distributed/_tensor/placement_types.py:
```diff
@@ -192,6 +192,8 @@ def _reduce_shard_tensor(
                 tensor, num_chunks, with_padding=True, contiguous=True
             )
             tensor = torch.cat(scattered_list, dim=self.dim)
+        else:
```
can you do 
```
elif not tensor.is_contiguous():
    tensor = tensor.contiguous()
```
instead?

wanchaol reviewed (commented):
Also please add some tests to test this behavior

yoyoyocmu commented:
The failed test is not related to this change, and seems like it became flaky since Dec 14: https://fburl.com/r3vmrgxb

wanchaol commented on test/distributed/test_device_mesh.py:
```diff
@@ -454,6 +454,46 @@ def test_all_gather_uneven(self):
+        # Partial to Shard to trigger reduce_scatter
+        partial_placement = _Partial()
+        shard_spec = Shard(0)
+        new_tensor = partial_placement._to_shard(
```
I think we should test the public API `from_local -> redistribute` instead, directly testing private APIs is fragile as it's not testing the user facing API and we might change private APIs frequently and introduces refactor burden

wanchaol reviewed (approved):
Some suggestions on the test, please address the comment before landing :)

yoyoyocmu committed:
Make input contiguous for DTensor reduce scatter to fix the incorrect numerical values (#115847)

Summary:

This change is to make the input tensor contiguous for DTensor reduce scatter in the case no padding is needed.

There's no exception thrown during training, but we ran into numerical value correctness issue without the change.

Test Plan:
**CI**
CI test

**WHEN model test**:
- Verified loss for each iteration within the expected range.
- Verified NE on-par with this change with 4B training data.

Reviewed By: wanchaol, Hahu803

Differential Revision: D52170822
